### DISCUSSION TITLE
Issue with encapsulated Arrays in Structure

### DISCUSSION THREAD

**Discussion by johannwesely** [eclipse-milo/milo#1022] [General]
![image](https://user-images.githubusercontent.com/108257325/180962527-18e598fc-0824-42ac-a057-c6768a16ebea.png)

Hello I have a komplex Structure so it's a little bit difficult to explain where the issue lies. On my milo Server which contains an Array of a  Structure (enevelopeStructType) I have Another Array Structure (trackStructType) wich is included into A Structure (chartStructType) wich is included into the envelopeStructType . I put an UA-Expert screen so maybe you can follow me :). So when i Init my top structure with dummy data. The First tracks child is filled correctly but the second tracks child shows empty data. The data is correct while debugging but when it comes to encoding i can't follow anymore where the data gets lost. ```
```
ExtensionObject[] xo = new ExtensionObject[UaStructure.length];

xo = ExtensionObject.encodeArray(getServer().getSerializationContext(), UaStructure); 
  
```
So maybe someone could check if he/she has the same behaviour with a similar encapsulated Array Structure. Thanks in Advance

**Comment by johannwesely** [comment]
i forgot to mention that the structure context is correctly filled maybe because its not an array Type.

**Comment by kevinherron** [comment]
All of the actual encoding is done by your codec implementations for each of the structures - can you share those?

**Comment by johannwesely** [comment]
```
	```
public static class Codec extends CodecSubClass implements OpcUaBinaryDataTypeCodec<ChartStructType> {
		@Override
		public Class<ChartStructType> getType() {
			return ChartStructType.class;
		}

		@Override
		public ChartStructType decode(SerializationContext context, OpcUaBinaryStreamDecoder reader) throws UaSerializationException {

			ContextStructType contextStructure = ReadStructureObject(f1, reader, new ContextStructType.Codec(), ContextStructType.class);
			TracksStructType[] tracksStructure = ReadStructureObjectArray(f2, reader, TracksStructType.TYPE_ID, TracksStructType[].class);

			return new ChartStructType(contextStructure, tracksStructure);
		}

		@Override
		public void encode(SerializationContext context, OpcUaBinaryStreamEncoder writer, ChartStructType value) throws UaSerializationException {

			writer.writeStruct(f1, value.Context, ContextStructType.TYPE_ID);
			writer.writeStructArray(f2, value.Tracks, TracksStructType.TYPE_ID);
		}
	}
```
```
....
public static class Codec extends CodecSubClass implements OpcUaBinaryDataTypeCodec<EnvelopeStructType> {
		@Override
		public Class<EnvelopeStructType> getType() {
			return EnvelopeStructType.class;
		}

		@Override
		public EnvelopeStructType decode(SerializationContext context, OpcUaBinaryStreamDecoder reader) throws UaSerializationException {

			Integer channel = reader.readInt32(f1);
			ChartStructType upperStopStructure = ReadStructureObject(f2, reader, new ChartStructType.Codec(), ChartStructType.class);
			ChartStructType lowerStopStructure = ReadStructureObject(f3, reader, new ChartStructType.Codec(), ChartStructType.class);

			return new EnvelopeStructType(channel, upperStopStructure, lowerStopStructure);
		}

		@Override
		public void encode(SerializationContext context, OpcUaBinaryStreamEncoder writer, EnvelopeStructType value) throws UaSerializationException {

			writer.writeInt32(f1, value.Channel);
			writer.writeStruct(f2, value.UpperStop, ChartStructType.TYPE_ID);
			writer.writeStruct(f3, value.LowerStop, ChartStructType.TYPE_ID);
		}
	}
....

```
```
public static class Codec extends CodecSubClass implements OpcUaBinaryDataTypeCodec<TracksStructType> {
		@Override
		public Class<TracksStructType> getType() {
			return TracksStructType.class;
		}

		@Override
		public TracksStructType decode(SerializationContext context, OpcUaBinaryStreamDecoder reader) throws UaSerializationException {

			Double[] doubleStructure = reader.readArray(f1, (value) -> reader.readDouble(f1), Double.class);
			String Quantity = reader.readString(f2);
			String Unit = reader.readString(f3);
			Integer Channel = reader.readInt32(f4);

			return new TracksStructType(doubleStructure, Quantity, Unit, Channel);
		}

		@Override
		public void encode(SerializationContext context, OpcUaBinaryStreamEncoder writer, TracksStructType value) throws UaSerializationException {
			writer.writeArray(f1, value.values, (val, valu) -> writer.writeDouble(valu));
			writer.writeString(f2, value.quantity);
			writer.writeString(f3, value.unit);
			writer.writeInt32(f4, value.channel);

		}
	}
```
```

**Comment by johannwesely** [comment]
With CodecSubClass

```
`public class CodecSubClass {
	
	
	public <T extends ToxUaStructure> T ReadStructureObject( String field, OpcUaBinaryStreamDecoder reader, DataTypeCodec codec, Class<?> t )
	{
	  Object object = reader.readStruct( field, codec );
	  T p1;
	  if (object.getClass().isAssignableFrom( t ) ) {
		p1 = ( T ) object;
	  }
	  else
		p1 = null;	
	  
	  return p1;
	}
	
	public <T extends ToxUaStructure> T[] ReadStructureObjectArray( String field, OpcUaBinaryStreamDecoder reader, ExpandedNodeId expandedNodeId, Class<?> t )
	{
	  Object object = reader.readStructArray( field, expandedNodeId );
	  T[] p1;
	  if (object.getClass().isAssignableFrom( t ) ) {
		p1 = ( T[] ) object;
	  }
	  else
		p1 = null;	
	  
	  return p1;
	}	
	
	
}`
```

**Reply by kevinherron** [reply]
It seems okay… without being able to fully reproduce and debug this the next obvious thing would be that you’re just not building your combination of structures correctly… have you set a breakpoint in the encoding of `ChartStructType` and verified you have a non-zero-length array to encode for the Tracks field each time?

**Comment by johannwesely** [comment]
![image](https://user-images.githubusercontent.com/108257325/181011455-519f91a7-834a-419d-949e-5f5b487fb4cf.png)

everytime i reach here the tracks array is non Zero-lenght.

**Reply by kevinherron** [reply]
You sure every time? Something seems off - in your picture the lengths of the Track field were 1 and 0, but here you have an array length 2.

**Reply by kevinherron** [reply]
Can you show the fully expanded debugger view for the top level envelope struct, showing all members fully expanded as well? What is the ContextStructType?

**Comment by johannwesely** [comment]
![image](https://user-images.githubusercontent.com/108257325/181018373-eee7ee9e-e613-4787-bc50-5101d2951b30.png)

Ah im sorry the debug picture is taken from a Structure (ChartStructType Array see Ua expert picture)  wich comes before. Funnily here the tracks structure works fine,  but when it comes to EnvelopeStructType i have the mentioned issue. But i double checked that the  tracks array is non Zero-lenght for all breaks.

**Comment by johannwesely** [comment]
this goes faster than i can type :) i apprechiate your help.
![image](https://user-images.githubusercontent.com/108257325/181020519-ad05fc11-1470-4466-991f-1e4be7ab14df.png)

**Comment by johannwesely** [comment]
![image](https://user-images.githubusercontent.com/108257325/181021470-d752d814-55d9-4a6e-9fb6-7161349d574f.png)
Here the fully expanded

**Reply by johannwesely** [reply]
![image](https://user-images.githubusercontent.com/108257325/181023698-68d56709-978e-4fe7-961b-c0e61eb53606.png)

**Comment by kevinherron** [comment]
Before any more hair pulling, can you see if Prosys OPC UA Browser also sees the same thing as UaExpert?

If it does, then my next guess would be maybe something going wrong encoding the `Context` field that precedes the `Tracks` arrays.

Or perhaps a mismatch in the definition you advertise for each structure vs.  how you actually encode it, causing decoders to end up at the wrong index in the stream.

**Comment by johannwesely** [comment]
Hi i used a Softing OpcUa Client with the same Results also we have our Own Milo Client wich shows the same Issue. Also i have checked the context Structure encoding wich seems to be correct. 
When i check the registered data Types i realize that EnvelopeStructType is registered with ArrayDimension: Null maybe thats the problem? 
![image](https://user-images.githubusercontent.com/108257325/181209800-d5263958-f993-40c1-9a17-ef57878f24bb.png)

But i dont know how to register my Structure as an array. Here is my Registration Method for Custom Structures 
```
protected void registerToxStructType(ToxUaStructure UaStructure) throws Exception {

		NodeId dataTypeId = UaStructure.getTypeId().toNodeIdOrThrow(getServer().getNamespaceTable());

		NodeId binaryEncodingId = UaStructure.getBinaryEncodingId().toNodeIdOrThrow(getServer().getNamespaceTable());

		String structureClassName = UaStructure.getClass().getSimpleName();

		dictionaryManager.registerStructureCodec(UaStructure.GetCodec(), structureClassName, dataTypeId, binaryEncodingId);

		StructureField[] fields = UaStructure.getStructureField(getServer());

		StructureDefinition definition = new StructureDefinition(binaryEncodingId, Identifiers.Structure, StructureType.Structure, fields);

		StructureDescription description = new StructureDescription(dataTypeId, new QualifiedName(getNamespaceIndex(), structureClassName), definition);

		dictionaryManager.registerStructureDescription(description, binaryEncodingId);
	}
```

**Reply by kevinherron** [reply]
The first time a client accesses the dictionary the server should log the temp file it's using to hold it, something like:
```
Wrote dictionary for '...' to '...'
```

It would be useful if you could attach the full dictionary here. What I was thinking is that one of your structure definitions doesn't match the codec behavior exactly.

**Comment by johannwesely** [comment]
```
<?xml version="1.0" encoding="UTF-8" standalone="true"?>

-<opc:TypeDictionary DefaultByteOrder="LittleEndian" TargetNamespace="urn:eclipse:tox:opcuaserver" xmlns:opc="http://opcfoundation.org/BinarySchema/">

<opc:Import Namespace="http://opcfoundation.org/BinarySchema/"/>

<opc:Import Namespace="urn:eclipse:tox:opcuaserver"/>


-<opc:StructuredType Name="SingleWindowResultStructType">

<opc:Field Name="Editable" TypeName="opc:Boolean"/>

<opc:Field Name="WindowNr" TypeName="opc:Int32"/>

<opc:Field Name="Text" TypeName="opc:String"/>

<opc:Field Name="QuantityX" TypeName="opc:String"/>

<opc:Field Name="QuantityY" TypeName="opc:String"/>

<opc:Field Name="QuantityZ" TypeName="opc:String"/>

<opc:Field Name="PointsLength" TypeName="opc:Int32"/>

<opc:Field Name="Points" TypeName="ns2:PointStructType" LengthField="PointsLength" xmlns:ns2="urn:eclipse:tox:opcuaserver"/>

<opc:Field Name="EdgeTypes" TypeName="opc:String"/>

<opc:Field Name="UniqeId" TypeName="opc:Int32"/>

<opc:Field Name="WindowIndex" TypeName="opc:Int32"/>

<opc:Field Name="EvaluationResult" TypeName="opc:Int32"/>

<opc:Field Name="Reason" TypeName="opc:Int32"/>

<opc:Field Name="Overall" TypeName="ns2:StateResultStructType" xmlns:ns2="urn:eclipse:tox:opcuaserver"/>

<opc:Field Name="SlopeMonitoring" TypeName="opc:Int32"/>

<opc:Field Name="TimeMonitoring" TypeName="opc:Int32"/>

<opc:Field Name="EdgeEvaluation" TypeName="ns2:StateResultStructType" xmlns:ns2="urn:eclipse:tox:opcuaserver"/>

<opc:Field Name="WindowEvaluation" TypeName="ns2:StateResultStructType" xmlns:ns2="urn:eclipse:tox:opcuaserver"/>

<opc:Field Name="ExtendedWindowEvaluation" TypeName="ns2:StateResultStructType" xmlns:ns2="urn:eclipse:tox:opcuaserver"/>

<opc:Field Name="AverageEvaluation" TypeName="opc:Int32"/>

<opc:Field Name="Severity" TypeName="opc:String"/>

<opc:Field Name="Entry" TypeName="ns2:PointStructType" xmlns:ns2="urn:eclipse:tox:opcuaserver"/>

<opc:Field Name="Leave" TypeName="ns2:PointStructType" xmlns:ns2="urn:eclipse:tox:opcuaserver"/>

<opc:Field Name="MaxByForce" TypeName="ns2:PointStructType" xmlns:ns2="urn:eclipse:tox:opcuaserver"/>

<opc:Field Name="MinByForce" TypeName="ns2:PointStructType" xmlns:ns2="urn:eclipse:tox:opcuaserver"/>

<opc:Field Name="AverageForce" TypeName="opc:Double"/>

<opc:Field Name="TimeSpent" TypeName="opc:Double"/>

<opc:Field Name="MaxSlope" TypeName="opc:Double"/>

<opc:Field Name="MinSlope" TypeName="opc:Double"/>

<opc:Field Name="AverageSlope" TypeName="opc:Double"/>

<opc:Field Name="MathFunction" TypeName="opc:Double"/>

<opc:Field Name="MathOutput1" TypeName="opc:Double"/>

<opc:Field Name="MathOutput2" TypeName="opc:Double"/>

<opc:Field Name="MathOutput3" TypeName="opc:Double"/>

<opc:Field Name="MathOutput4" TypeName="opc:Double"/>

<opc:Field Name="ResultingOutput1" TypeName="opc:Int32"/>

<opc:Field Name="ResultingOutput2" TypeName="opc:Int32"/>

<opc:Field Name="ResultingOutput3" TypeName="opc:Int32"/>

<opc:Field Name="ResultingOutput4" TypeName="opc:Int32"/>

<opc:Field Name="ValueOfCalcFunctionEvaluation" TypeName="opc:Double"/>

<opc:Field Name="ResultOfCalcFunctionEvaluation" TypeName="ns2:StateResultStructType" xmlns:ns2="urn:eclipse:tox:opcuaserver"/>

<opc:Field Name="EvaluationTarget" TypeName="opc:Int32"/>

<opc:Field Name="CalculationFunction" TypeName="opc:Int32"/>

<opc:Field Name="MinMaxMeanCalcDefinitions" TypeName="opc:Int32"/>

<opc:Field Name="WindowExtendedLines" TypeName="opc:Int32"/>

</opc:StructuredType>


-<opc:StructuredType Name="ContextStructType">

<opc:Field Name="EnvelopeMonitored" TypeName="opc:Boolean"/>

<opc:Field Name="EnvelopeWarningMonitored" TypeName="opc:Boolean"/>

<opc:Field Name="TeachActive" TypeName="opc:Boolean"/>

<opc:Field Name="EnvelopeErrorLower" TypeName="opc:Boolean"/>

<opc:Field Name="EnvelopeErrorUpper" TypeName="opc:Boolean"/>

<opc:Field Name="EnvelopeWarningLower" TypeName="opc:Boolean"/>

<opc:Field Name="EnvelopeWarningUpper" TypeName="opc:Boolean"/>

<opc:Field Name="PointOfEnvelopeViolation" TypeName="ns2:XYPointStruct" xmlns:ns2="urn:eclipse:tox:opcuaserver"/>

<opc:Field Name="PointOfWarningEnvelopeViolation" TypeName="ns2:XYPointStruct" xmlns:ns2="urn:eclipse:tox:opcuaserver"/>

<opc:Field Name="IndexOfEnvelopeViolation" TypeName="opc:Int32"/>

<opc:Field Name="IndexOfWarningEnvelopeViolation" TypeName="opc:Int32"/>

</opc:StructuredType>


-<opc:StructuredType Name="SubTypesMapStructType">

<opc:Field Name="Index" TypeName="opc:Int32"/>

<opc:Field Name="SubType" TypeName="opc:String"/>

</opc:StructuredType>


-<opc:StructuredType Name="MeasurementStructType">

<opc:Field Name="X" TypeName="ns2:MinMaxActualValueStructType" xmlns:ns2="urn:eclipse:tox:opcuaserver"/>

<opc:Field Name="Y" TypeName="ns2:MinMaxActualValueStructType" xmlns:ns2="urn:eclipse:tox:opcuaserver"/>

<opc:Field Name="Z" TypeName="ns2:MinMaxActualValueStructType" xmlns:ns2="urn:eclipse:tox:opcuaserver"/>

</opc:StructuredType>


-<opc:StructuredType Name="PropertiesAndSubtypeStructType">

<opc:Field Name="TypeName" TypeName="opc:String"/>

<opc:Field Name="Version" TypeName="opc:Int32"/>

<opc:Field Name="PropertiesMap" TypeName="ns2:PropertiesMapStructType" xmlns:ns2="urn:eclipse:tox:opcuaserver"/>

<opc:Field Name="SubTypesMap" TypeName="ns2:SubTypesMapStructType" xmlns:ns2="urn:eclipse:tox:opcuaserver"/>

</opc:StructuredType>


-<opc:StructuredType Name="MinMaxActualValueStructType">

<opc:Field Name="Min" TypeName="opc:Double"/>

<opc:Field Name="Max" TypeName="opc:Double"/>

<opc:Field Name="Actual" TypeName="opc:Double"/>

<opc:Field Name="Unit" TypeName="opc:String"/>

</opc:StructuredType>


-<opc:StructuredType Name="XYPointStruct">

<opc:Field Name="X" TypeName="opc:Double"/>

<opc:Field Name="Y" TypeName="opc:Double"/>

</opc:StructuredType>


-<opc:StructuredType Name="ChartStructType">

<opc:Field Name="Context" TypeName="ns2:ContextStructType" xmlns:ns2="urn:eclipse:tox:opcuaserver"/>

<opc:Field Name="TracksLength" TypeName="opc:Int32"/>

<opc:Field Name="Tracks" TypeName="ns2:TracksStructType" LengthField="TracksLength" xmlns:ns2="urn:eclipse:tox:opcuaserver"/>

</opc:StructuredType>


-<opc:StructuredType Name="PropertiesMapStructType">

<opc:Field Name="Key" TypeName="opc:String"/>

<opc:Field Name="Value" TypeName="opc:String"/>

</opc:StructuredType>


-<opc:StructuredType Name="ChannelWindowResultStructType">

<opc:Field Name="Channel" TypeName="opc:Int32"/>

<opc:Field Name="SingleWindowResultLength" TypeName="opc:Int32"/>

<opc:Field Name="SingleWindowResult" TypeName="ns2:SingleWindowResultStructType" LengthField="SingleWindowResultLength" xmlns:ns2="urn:eclipse:tox:opcuaserver"/>

</opc:StructuredType>


-<opc:StructuredType Name="TracksStructType">

<opc:Field Name="Values" TypeName="opc:Double"/>

<opc:Field Name="Quantity" TypeName="opc:String"/>

<opc:Field Name="Unit" TypeName="opc:String"/>

<opc:Field Name="Channel" TypeName="opc:Int32"/>

</opc:StructuredType>


-<opc:StructuredType Name="StateResultStructType">

<opc:Field Name="ActivationState" TypeName="opc:Int32"/>

<opc:Field Name="Result" TypeName="opc:Int32"/>

</opc:StructuredType>


-<opc:StructuredType Name="EnvelopeStructType">

<opc:Field Name="Channel" TypeName="opc:Int32"/>

<opc:Field Name="UpperStop" TypeName="ns2:ChartStructType" xmlns:ns2="urn:eclipse:tox:opcuaserver"/>

<opc:Field Name="LowerStop" TypeName="ns2:ChartStructType" xmlns:ns2="urn:eclipse:tox:opcuaserver"/>

</opc:StructuredType>


-<opc:StructuredType Name="PointStructType">

<opc:Field Name="XVal" TypeName="opc:Double"/>

<opc:Field Name="YVal" TypeName="opc:Double"/>

<opc:Field Name="TimeVal" TypeName="opc:Double"/>

</opc:StructuredType>

</opc:TypeDictionary>
```

**Reply by kevinherron** [reply]
Something looks wrong with the definition for `TracksStructType`:
```
<opc:StructuredType Name="TracksStructType">
  <opc:Field Name="Values" TypeName="opc:Double"/>
  <opc:Field Name="Quantity" TypeName="opc:String"/>
  <opc:Field Name="Unit" TypeName="opc:String"/>
  <opc:Field Name="Channel" TypeName="opc:Int32"/>
</opc:StructuredType>
```

In your codec implementation you encode/decode "Values" as an array.

**Comment by johannwesely** [comment]
You are the Hero of my Day !!! Thanks a lot
 i had to change the StructureField of Values in  TracksStructType  from Double to Double Array now it works.
